enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,549 @@
+implementations to label all OpenShift components not running in
+static pods with the `io.openshift.management: true` label, and add a
+CI job to require that label be present in all workloads created from
+the release payload.
+
+We need kubelet to know when the feature is enabled, but we cannot
+change the configuration schema that kubelet gets from
+upstream. Therefore we will have kubelet look for a new configuration
+file on startup, and the feature will only be enabled if that file is
+found.
+
+We want to give cluster administrators control over which workloads
+are run on the management CPUs. Therefore, only pods in namespaces
+labeled with `io.openshift.management: true` will be subject to
+special handling. Normal users cannot add a label to a namespace
+without the right RBAC permissions.
+
+We want to continue to use the scheduler for placing management
+workloads, but we cannot rely on the CPU requests of those workloads
+to accurately reflect the constrained environment in which they are
+expected to run. Instead of scaling those CPU request values, we will
+change them to request a new [extended
+resource](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/),
+"management cores." We will modify kubelet to advertise management
+cores as an extended resource when the management workload
+partitioning feature is enabled, using a value equivalent to the CPU
+resources for the entire host. This large value should allow the
+scheduler to always be able to place workloads, while still accurately
+accounting for those requests.
+
+We need to ensure fair allocation of CPU time between different
+management workloads. Therefore, we will copy the original CPU
+requests for management workload pods into an annotation that CRI-O
+can use to configure the CPU shares when running the containers.
+
+We need to change pod definitions as each pod is created, so that the
+scheduler, kubelet, and CRI-O all see a consistently updated version
+of the pod and do not need to make independent decisions about whether
+it should be treated as a management workload. We need to intercept
+pod creation for *all* pods, without race conditions that might be
+introduced with traditional admission webhooks or controllers.
+Therefore, we will build an admission hook into the [kubernetes API
+server in
+OpenShift](https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission),
+to intercept API requests to create pods.
+
+The API server requires a configuration resource as input, rather than
+a ConfigMap or command line flag.  Therefore, we need an API-driven
+way to enable management workload partitioning in the admission
+hook. We will extend the Infrastructure configuration resource with a
+new status field to indicate whether the feature is on or off.
+
+*TBD: API field name, see open questions*
+
+Some pods used to run OpenShift control plane components are started
```

Comment:
can we enumerate?  static pod was the example i had in mind which is why kubelet itself needs to know how to do this conversion when reading static pods from disk prior to mirroring the pod back to apiserver.  this is why the kubelet carry is needed on static pod file parsing where it applies defaults today.

Reply:
The ones we identified are kubeapiserver, kube-controller-manager, scheduler, etcd. I will update the doc at this point to add the list.

Reply:
ok, all of those should have the label, and the kubelet admission check to strip/convert on static pods would cover the gap i believe.